examples/cascadeproxy-socks/socksproxy.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"crypto/tls"
+	"flag"
+	"github.com/elazarl/goproxy"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+type SocksAuth struct {
+	Username, Password string
+}
+
+func createSocksProxy(socksAddr string, auth SocksAuth) func(r *http.Request) (*url.URL, error) {
+	return func(r *http.Request) (*url.URL, error) {
+		Url := &url.URL{
+			Scheme: "socks5",
+			Host:   socksAddr,
+		}
+		if auth.Username != "" {
+			Url.User = url.UserPassword(auth.Username, auth.Password)
+		}
+		return Url, nil
+	}
+}
+
+func main() {
+	verbose := flag.Bool("v", false, "should every proxy request be logged to stdout")
+	addr := flag.String("addr", ":8080", "proxy listen address")
+	socksAddr := flag.String("socks", "127.0.0.1:1080", "socks proxy address")
+	username := flag.String("user", "", "username for SOCKS5 proxy if auth is required")
+	password := flag.String("pass", "", "password for SOCKS5 proxy")
+	flag.Parse()
+
+	auth := SocksAuth{
+		Username: *username,
+		Password: *password,
+	}
+	proxyServer := goproxy.NewProxyHttpServer()
+
+	proxyServer.OnRequest().Do(goproxy.FuncReqHandler(func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
+		resolvedAddr, _ := net.ResolveIPAddr("ip", req.URL.Host)
+		req.URL.Host = resolvedAddr.String() + ":" + req.URL.Port()
+		return req, nil
+	}))
+
+	proxyServer.OnRequest().HandleConnect(goproxy.FuncHttpsHandler(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
+		req := ctx.Req
+		req.RequestURI = ""
+		client := &http.Client{
+			Transport: &http.Transport{
+				Proxy: createSocksProxy(*socksAddr, auth),
```

Comment:
All this logic should be inside the Do / DoFunc function!
The connect handler must be only `goproxy.AlwaysMitm`.
This way the proxy server will take over the connection.

Then, in the Do function, you return the resulting *http.Response as the second parameter, and the server will return it directly to the client.

Reply:
Im having this error:

```go
	proxyServer.OnRequest().HandleConnect(goproxy.FuncHttpsHandler(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
		return &goproxy.ConnectAction{
			Action: goproxy.ConnectMitm,
		}, host
	}))
```

```
2024/12/20 20:25:20 [001] INFO: Running 1 CONNECT handlers
2024/12/20 20:25:20 [001] INFO: on 0th handler: &{2 <nil> <nil>} www.google.com:443
2024/12/20 20:25:20 [001] INFO: Assuming CONNECT is TLS, mitm proxying it
2024/12/20 20:25:20 [001] WARN: Cannot handshake client www.google.com:443 tls: no certificates configured
```

Its immediately  stop when its do the handshake

```go
	case ConnectMitm:
		proxyClient.Write([]byte("HTTP/1.0 200 OK\r\n\r\n"))
		ctx.Logf("Assuming CONNECT is TLS, mitm proxying it")
		// this goes in a separate goroutine, so that the net/http server won't think we're
		// still handling the request even after hijacking the connection. Those HTTP CONNECT
		// request can take forever, and the server will be stuck when "closed".
		// TODO: Allow Server.Close() mechanism to shut down this connection as nicely as possible
		tlsConfig := defaultTLSConfig
		if todo.TLSConfig != nil {
			var err error
			tlsConfig, err = todo.TLSConfig(host, ctx)
			if err != nil {
				httpError(proxyClient, ctx, err)
				return
			}
		}
		go func() {
			//TODO: cache connections to the remote website
			rawClientTls := tls.Server(proxyClient, tlsConfig)
			defer rawClientTls.Close()
			if err := rawClientTls.Handshake(); err != nil {
				ctx.Warnf("Cannot handshake client %v %v", r.Host, err)
				return
			}
```

Reply:
Instead of
```
&goproxy.ConnectAction{
			Action: goproxy.ConnectMitm,
		}
```

Please use
```
goproxy.AlwaysMitm
```

Reply:
I solved it with goproxy.AlwaysMitm but I encountered another issue:

```
curl: (60) SSL certificate problem: self-signed certificate in certificate chain
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

Reply:
The alternative to -k is the guide to trust the certificates, but no one wrote it yet :)
https://github.com/elazarl/goproxy/issues/560